Add polaris spark client webpage under unreleased (#1503)
apache/polaris: site/content/in-dev/unreleased/polaris-spark-client.md

---
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
Title: Polaris Spark Client
type: docs
weight: 400
---

Apache Polaris now provides Catalog support for Generic Tables (non-iceberg tables), please check out 
the [Catalog API Spec]({{% ref "polaris-catalog-service" %}}) for Generic Table API specs.

> Along with the Generic Table Catalog support, Polaris is also releasing a Spark Client, which help
> providing an end-to-end solution for Apache Spark to manage Delta tables using Polaris.
> 

Reviewer: ```suggestion
Along with the Generic Table Catalog support, Polaris is also releasing a Spark Client, which helps
provide an end-to-end solution for Apache Spark to manage Delta tables using Polaris.

```

Author: updated

Note the Polaris Spark Client is able to handle both Iceberg and Delta tables, not just Delta.

> This pages documents how to build and use the Apache Polaris Spark Client before formal release.

Reviewer: ```suggestion
This page documents how to build and use the Polaris Spark Client directly from soruce.
```

Author: updated

## Prerequisite
1. Check out the polaris repo
```shell
cd ~
git clone https://github.com/apache/polaris.git
```
2. Spark with version >= 3.5.3 and <= 3.5.5, recommended with 3.5.5.
```shell
cd ~
wget https://archive.apache.org/dist/spark/spark-3.5.5/spark-3.5.5-bin-hadoop3.tgz 
mkdir spark-3.5
tar xzvf spark-3.5.5-bin-hadoop3.tgz  -C spark-3.5 --strip-components=1
cd spark-3.5
```

All Spark Client code is available under `plugins/spark` of the polaris repo.

## Quick Start with Local Polaris Service
If you want to quickly try out the functionality with a local Polaris service, you can follow the instructions
in `plugins/spark/v3.5/getting-started/README.md`.

The getting-started will start two containers:
1) The `polaris` service for running Apache Polaris using an in-memory metastore
2) The `jupyter` service for running Jupyter notebook with PySpark (Spark 3.5.5 is used)

The notebook `SparkPolaris.ipynb` provided under `plugins/spark/v3.5/getting-started/notebooks` provides examples 
with basic commands, includes:

> 1) Connect to Polaris using Python client to create Catalog and Roles

Reviewer: ```suggestion
1) Connect to Polaris using Python client to create a Catalog and Roles
```

Author: updated

2) Start Spark session using the Polaris Spark Client
3) Using Spark to perform table operations for both Delta and Iceberg

## Start Spark against a deployed Polaris Service
If you want to start Spark with a deployed Polaris service, you can follow the following instructions.

> Before start, Make sure the service deployed is up-to-date, and Spark 3.5 with at least version 3.5.3 is installed. 

Reviewer: ```suggestion
Before starting, make sure the service deployed is up-to-date, and that Spark 3.5 with at least version 3.5.3 is installed. 
```

Author: updated

--conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,io.delta.sql.DeltaSparkSessionExtension \
--conf spark.sql.catalog.spark_catalog=org.apache.spark.sql.delta.catalog.DeltaCatalog \
--conf spark.sql.catalog.<spark-catalog-name>.warehouse=<polaris-catalog-name> \
--conf spark.sql.catalog.<spark-catalog-name>.header.X-Iceberg-Access-Delegation=vended-credentials \
--conf spark.sql.catalog.<spark-catalog-name>=org.apache.polaris.spark.SparkCatalog \
--conf spark.sql.catalog.<spark-catalog-name>.uri=<polaris-service-uri> \
--conf spark.sql.catalog.<spark-catalog-name>.credential='<client-id>:<client-secret>' \
--conf spark.sql.catalog.<spark-catalog-name>.scope='PRINCIPAL_ROLE:ALL' \
--conf spark.sql.catalog.<spark-catalog-name>.token-refresh-enabled=true
```

Replace `path-to-spark-client-jar` to where the built jar is located. The `spark-catalog-name` is the catalog name you
will use with spark, and `polaris-catalog-name` is the catalog name used by Polaris service, for simplicity, you can use
the same name. Replace the `polaris-service-uri`, `client-id` and `client-secret` accordingly, you can refer to

> [Quick Start]({{% ref "../0.9.0/quickstart" %}}) for more details about those fields.

Reviewer: doubt: should this really link to 0.9.0?

Author: right now, seems we only have 0.9.0, we haven't update the code to have 0.10.0 yet, but i felt we should have some page that is not version related to describe the basic concept that does not change across version

Reviewer: I think there's an unreleased page, can we use that?

Or you can create a spark session start the connection, following is an example with pyspark

> ```shell

Reviewer: ```suggestion
```python
```

  .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")
  .config("spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,io.delta.sql.DeltaSparkSessionExtension")
  .config("spark.sql.catalog.<spark-catalog-name>", "org.apache.polaris.spark.SparkCatalog")  
  .config("spark.sql.catalog.<spark-catalog-name>.uri", <polaris-service-uri>)
  .config("spark.sql.catalog.<spark-catalog-name>.token-refresh-enabled", "true")
  .config("spark.sql.catalog.<spark-catalog-name>.credential", "<client-id>:<client_secret>")
  .config("spark.sql.catalog.<spark-catalog-name>.warehouse", <polaris_catalog_name>)
  .config("spark.sql.catalog.polaris.scope", 'PRINCIPAL_ROLE:ALL')
  .config("spark.sql.catalog.polaris.header.X-Iceberg-Access-Delegation", 'vended-credentials')
  .getOrCreate()
```
Similar as the CLI command, make sure the corresponding fields are replaced correctly.

### Create tables with Spark
After the spark is started, you can use it to create and access Iceberg and Delta table like what you are doing before, 
for example:

> ```shell

Reviewer: ```suggestion
```python
```

Author: updated

spark.sql("USE polaris")
spark.sql("CREATE NAMESPACE IF NOT EXISTS DELTA_NS")
spark.sql("CREATE NAMESPACE IF NOT EXISTS DELTA_NS.PUBLIC")
spark.sql("USE NAMESPACE DELTA_NS.PUBLIC")
spark.sql("""CREATE TABLE IF NOT EXISTS PEOPLE (
    id int, name string)
USING delta LOCATION 'file:///tmp/delta_tables/people';
""")
```

## Limitations
The Polaris Spark client has the following functionality limitations:
1) Create table as select (CTAS) is not supported for Delta tables. As a result, the `saveAsTable` method of `Dataframe`
   is also not supported, since it relies on the CTAS support.
2) Create a Delta table without explicit location is not supported.
3) Rename a Delta table is not supported.
4) ALTER TABLE ... SET LOCATION/SET FILEFORMAT/ADD PARTITION is not supported for DELTA table.

> 5) For other non-iceberg tables like csv, there is no specific guarantee provided today.

Reviewer: ```suggestion
5) For other non-Iceberg tables like csv, there is no specific guarantee provided today.
```

Author: updated

> 6) Role-based RBAC support for Delta table write is not available. Create, Drop and List RBAC support is available.

Reviewer: What exactly does this mean? Can we just say credential vending isn't supported? Technically the RB in RBAC already stands for Role-based

Author: All i mean is TABLE_WRITE_DATA privilege is currently not supported. i updated to make it more specific, and I also added that credential vending is not working

# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
Title: Polaris Spark Client
type: docs
weight: 400
---

Apache Polaris now provides Catalog support for Generic Tables (non-iceberg tables), please check out 
the [Catalog API Spec]({{% ref "polaris-catalog-service" %}}) for Generic Table API specs.

Along with the Generic Table Catalog support, Polaris is also releasing a Spark Client, which helps to 
provide an end-to-end solution for Apache Spark to manage Delta tables using Polaris.

> Note the Polaris Spark Client is able to handle both Iceberg and Delta tables, not just Delta.

Reviewer 2: Q: do parquet/csv tables work?

Author: I briefly tested scv manually before, basic operations like create, insert, drop works, alter doesn't work. I haven't tested parquet yet. 
I don't want to do specific commitment for those table formats yet before we do more detailed testing.

Reviewer: We should probably file an issue for this

> This page documents how to build and use the Polaris Spark Client directly with the source repo.

Reviewer 2: Can we talk about how to use the jar, then goes to build details? I believe most of users mainly care about the usage once we release the client jar.

Author: I adjusted the order, and have section 
## Start Spark against a deployed Polaris service
and 
## Connecting with Spark using local Polaris Spark client

## Prerequisite
1. Check out the polaris repo
```shell
cd ~
git clone https://github.com/apache/polaris.git
```

> 2. Spark with version >= 3.5.3 and <= 3.5.5, recommended with 3.5.5.

Reviewer 2: ```suggestion
2. Download Spark distributioni with version >= 3.5.3 and <= 3.5.5, recommended with 3.5.5.
```

Author: This is not needed for Quick Start, so i moved it under 'Start Spark against a deployed Polaris service' now

---
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
Title: Polaris Spark Client
type: docs

> weight: 400

Reviewer: +1 to a new folder

> weight: 400

Reviewer 2: We could create a new folder once we got more than one client, e.g. spark and Trino clients.

> weight: 400

Author: I think a new folder seems to much for now, i moved after Deploying in Production, since the spark client can only be used after the polaris is deployed

---

Apache Polaris now provides Catalog support for Generic Tables (non-Iceberg tables), please check out 
the [Catalog API Spec]({{% ref "polaris-catalog-service" %}}) for Generic Table API specs.

Along with the Generic Table Catalog support, Polaris is also releasing a Spark client, which helps to 
provide an end-to-end solution for Apache Spark to manage Delta tables using Polaris.

Note the Polaris Spark client is able to handle both Iceberg and Delta tables, not just Delta.

This page documents how to connect Spark with Polaris Service using the Polaris Spark client.

## Prerequisite
Check out the Polaris repo:
```shell
cd ~
git clone https://github.com/apache/polaris.git
```

> All Spark client code is available under `plugins/spark` of the Polaris repo.

Author: That is also needed for the quick start, but since I removed most of the instruction for getting started and local Polaris Spark client, i moved this instruction to quick start

## Quick Start with Local Polaris service
If you want to quickly try out the functionality with a local Polaris service, you can follow the instructions
in `plugins/spark/v3.5/getting-started/README.md`.

The getting-started will start two containers:
1) The `polaris` service for running Apache Polaris using an in-memory metastore
2) The `jupyter` service for running Jupyter notebook with PySpark (Spark 3.5.5 is used)

The notebook `SparkPolaris.ipynb` provided under `plugins/spark/v3.5/getting-started/notebooks` provides examples 
with basic commands, includes:
1) Connect to Polaris using Python client to create a Catalog and Roles
2) Start Spark session using the Polaris Spark client

> 3) Using Spark to perform table operations for both Delta and Iceberg

Author: good point, i removed most of the description here, and added a link to the README

## Start Spark against a deployed Polaris service

> Before starting, make sure the service deployed is up-to-date, and that Spark 3.5 with at least version 3.5.3 is installed.

Author: updated

Spark 3.5.5 is recommended, and you can follow the instructions below to get a Spark 3.5.5 distribution.
```shell
cd ~
wget https://archive.apache.org/dist/spark/spark-3.5.5/spark-3.5.5-bin-hadoop3.tgz 
mkdir spark-3.5
tar xzvf spark-3.5.5-bin-hadoop3.tgz  -C spark-3.5 --strip-components=1
cd spark-3.5
```

### Connecting with Spark using the Polaris Spark client

> The following CLI command can be used to start the spark with connection to the deployed Polaris service using

Reviewer 3: "start Apache Spark with a connection"

Author: Capitalized the Spark, since I have Apache Spark at the very beginning, i don't think we need to repeat Apache Spark everywhere

--conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,io.delta.sql.DeltaSparkSessionExtension \
--conf spark.sql.catalog.spark_catalog=org.apache.spark.sql.delta.catalog.DeltaCatalog \
--conf spark.sql.catalog.<spark-catalog-name>.warehouse=<polaris-catalog-name> \
--conf spark.sql.catalog.<spark-catalog-name>.header.X-Iceberg-Access-Delegation=vended-credentials \
--conf spark.sql.catalog.<spark-catalog-name>=org.apache.polaris.spark.SparkCatalog \
--conf spark.sql.catalog.<spark-catalog-name>.uri=<polaris-service-uri> \
--conf spark.sql.catalog.<spark-catalog-name>.credential='<client-id>:<client-secret>' \
--conf spark.sql.catalog.<spark-catalog-name>.scope='PRINCIPAL_ROLE:ALL' \
--conf spark.sql.catalog.<spark-catalog-name>.token-refresh-enabled=true
```
Assume the released Polaris Spark client you want to use is `org.apache.polaris:polaris-iceberg-1.8.1-spark-runtime-3.5_2.12:1.0.0`,
replace the `polaris-spark-client-package` field with the release.

The `spark-catalog-name` is the catalog name you will use with Spark, and `polaris-catalog-name` is the catalog name used 
by Polaris service, for simplicity, you can use the same name. 

> Replace the `polaris-service-uri`, `client-id` and `client-secret` accordingly, you can refer to

Reviewer 3: `polaris-service-uri` is not described in "Using Polaris" - you should state what the correct value should be here based on a standard local Spark deployment.

[Using Polaris]({{% ref "getting-started/using-polaris" %}}) for more details about those fields.

> You can also start the connection by creating a Spark session, following is an example with PySpark:

Reviewer 3: > You can also start the connection by creating a Spark session

Isn't the above example also creating a Spark session?

Reviewer: I think what this is trying to say is that you can add the catalog from within a Spark session that's already started

Author: What i really mean is just programmatically initialize a SparkSession not through a command line, updated

```python
from pyspark.sql import SparkSession

spark = SparkSession.builder
  .config("spark.jars.packages", "<polaris-spark-client-package>,org.apache.hadoop:hadoop-aws:3.3.4,io.delta:delta-spark_2.12:3.3.1")
  .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")
  .config("spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,io.delta.sql.DeltaSparkSessionExtension")
  .config("spark.sql.catalog.<spark-catalog-name>", "org.apache.polaris.spark.SparkCatalog")  
  .config("spark.sql.catalog.<spark-catalog-name>.uri", <polaris-service-uri>)
  .config("spark.sql.catalog.<spark-catalog-name>.token-refresh-enabled", "true")
  .config("spark.sql.catalog.<spark-catalog-name>.credential", "<client-id>:<client_secret>")
  .config("spark.sql.catalog.<spark-catalog-name>.warehouse", <polaris_catalog_name>)
  .config("spark.sql.catalog.polaris.scope", 'PRINCIPAL_ROLE:ALL')
  .config("spark.sql.catalog.polaris.header.X-Iceberg-Access-Delegation", 'vended-credentials')
  .getOrCreate()
```
Similar as the CLI command, make sure the corresponding fields are replaced correctly.

### Create tables with Spark

> After the Spark is started, you can use it to create and access Iceberg and Delta table like what you are doing before,

Reviewer 3: "After Apache Spark is started"

Remove: "like what you are doing before" - I don't think this sentence fragment adds to the overall message.

Reviewer: You don't need `Apache` everywhere; there are already preceding usages of Spark without it

> After the Spark is started, you can use it to create and access Iceberg and Delta table like what you are doing before,

Reviewer: ```suggestion
After Spark is started, you can use it to create and access Iceberg and Delta tables,
```

Author: updated

for example:
```python
spark.sql("USE polaris")
spark.sql("CREATE NAMESPACE IF NOT EXISTS DELTA_NS")
spark.sql("CREATE NAMESPACE IF NOT EXISTS DELTA_NS.PUBLIC")
spark.sql("USE NAMESPACE DELTA_NS.PUBLIC")
spark.sql("""CREATE TABLE IF NOT EXISTS PEOPLE (
    id int, name string)

> USING delta LOCATION 'file:///tmp/delta_tables/people';

Reviewer 3: In general, we've switched to using "/var/tmp/" instead of "/tmp/" in Getting Started due to tmp dir's GC sometimes being quick to dump

    id int, name string)
USING delta LOCATION 'file:///tmp/var/delta_tables/people';
""")
```

> ## Connecting with Spark using local Polaris Spark client

Author: sg! I removed most of it and added a link to the README

> If there is no released Spark client, or you want to try the Spark client that is currently not yet released. You can
> build a Spark Client jar locally with the source repo, and use the local jar to connect Spark with Polaris Service.

Reviewer: ```suggestion
If you'd like to use version of the Spark client that is currently not yet released you can
build a Spark Client jar locally with the source repo, and use the local jar to connect Spark with Polaris Service.
```

Author: uppdated

--conf spark.sql.catalog.<spark-catalog-name>=org.apache.polaris.spark.SparkCatalog \
--conf spark.sql.catalog.<spark-catalog-name>.uri=<polaris-service-uri> \
--conf spark.sql.catalog.<spark-catalog-name>.credential='<client-id>:<client-secret>' \
--conf spark.sql.catalog.<spark-catalog-name>.scope='PRINCIPAL_ROLE:ALL' \
--conf spark.sql.catalog.<spark-catalog-name>.token-refresh-enabled=true
```
Replace `path-to-spark-client-jar` with where the built jar is located.

## Limitations
The Polaris Spark client has the following functionality limitations:
1) Create table as select (CTAS) is not supported for Delta tables. As a result, the `saveAsTable` method of `Dataframe`
   is also not supported, since it relies on the CTAS support.
2) Create a Delta table without explicit location is not supported.
3) Rename a Delta table is not supported.
4) ALTER TABLE ... SET LOCATION/SET FILEFORMAT/ADD PARTITION is not supported for DELTA table.
5) For other non-Iceberg tables like csv, there is no specific guarantee provided today.
6) TABLE_WRITE_DATA privilege is not supported for Delta Table.

> 7) Credential Vending is not supported for Delta Table.

Reviewer 3: Same question as above.
